lib/Dialect/ATen/ATenLoweringPass.cpp

```diff
@@ -0,0 +1,972 @@
+};
+
+/// Lower ReLU
+class ReLUOpConversion : public ConversionPattern {
+public:
+  explicit ReLUOpConversion(MLIRContext *context)
+      : ConversionPattern(mlir::NPCOMP::aten::ReluOp::getOperationName(), 1,
+                          context) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    return rewriteWithFunctionCall(op, operands, rewriter, "relu");
+  }
+};
+
+/// Lower ThresholdBackward
+class ThresholdBackwardOpConversion : public ConversionPattern {
+public:
+  explicit ThresholdBackwardOpConversion(MLIRContext *context)
+      : ConversionPattern(mlir::NPCOMP::aten::ThresholdBackwardOp::getOperationName(),
+                          1, context) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    return rewriteWithFunctionCall(op, operands, rewriter, "threshold_backward");
+  }
+};
+
+/// Lower transpose
+class TransposeOpConversion : public ConversionPattern {
+public:
+  explicit TransposeOpConversion(MLIRContext *context)
+      : ConversionPattern(mlir::NPCOMP::aten::TOp::getOperationName(), 1, context) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    return rewriteWithFunctionCall(op, operands, rewriter, "t");
+  }
+};
+
+/// Lower view
+class ViewOpConversion : public ConversionPattern {
+public:
+  explicit ViewOpConversion(MLIRContext *context)
+      : ConversionPattern(mlir::NPCOMP::aten::ViewOp::getOperationName(), 1,
+                          context) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto loc = op->getLoc();
+    edsc::ScopedContext scope(rewriter, loc);
+
+    Value xVal = MemRefTypeCast(rewriter, operands[0]);
+
+     // construct the shape argument
+    SmallVector<Value, 8> shape;
+    auto co = dyn_cast<mlir::NPCOMP::aten::ConstantOp>(operands[1].getDefiningOp());
+    DenseElementsAttr a = co.template getAttrOfType<DenseElementsAttr>("value");
+    for (auto i : a.getAttributeValues())
+      shape.push_back(rewriter.create<mlir::ConstantOp>(co.getLoc(), i));
+
+
+    // pad out the shape with -1 to make it 4d
+    while (shape.size() < 4)
+      shape.push_back(constInt(-1, 32));
+
+    SmallVector<Value, 8> callops{xVal, shape[0], shape[1], shape[2], shape[3]};
+
+    return rewriteWithVoidFunctionCallExplicit(op, callops, operands, rewriter, "view");
+  }
+};
+
+/// Convert an ATen type, this gets called for block and region arguments, and
+/// attributes.
+MemRefType convertTensorType(TensorType tensor) {
+  return mlir::MemRefType::get(tensor.getShape(), tensor.getElementType(), {},
+                               0);
+}
+
+struct ATenLoweringPass
```

Comment:
Would you mind adding a comment describing the invariants, and conversions that are supported (and what is currently not supported).

Reply:
Moved to issue